Quote

DenisKarch opened the pull request:
Added Quote
made wellKnown global

ericchiang commented on attestation/attestation.go:
```diff
@@ -22,6 +22,8 @@ import (
 	"github.com/google/go-tspi/tspiconst"
 )
 
+var wellKnown [20]byte
```
I know you're just moving this, but since it's global now can you rename it to something more descriptive. E.g. 'wellKnownSecret' or whatever it's being used as. Also add a comment

ericchiang commented on attestation/attestation.go:
```diff
@@ -141,10 +140,52 @@ func CreateAIK(context *tspi.Context) ([]byte, []byte, error) {
+		return nil, nil, fmt.Errorf("failed to get a reference to PCRs: %v", err)
+	}
+
+	bitmap := make([]int, 24)
```
Why is this created but not returned? Please add a comment and explain what the values you're setting it to.

DenisKarch replied:
I suppose it's a bit inaccurate to call it a bitmap. SetPCRs consumes a slice of chosen PCRs (when we later get the PCR values it will only fetch values of PCRs in the slice). I would much prefer to do something like pcrs.SetPCRS(range(24)) but I don't believe there is anyway to do so in go?

ericchiang replied:
Try this:

```
pcrValues := make([]int, 24)
for i := range pcrValues {
    pcrValues[i] = i
}
if err := pcrs.SetPCRs(pcrValues); err != nil {
    // ...
}
```

Also please be more descriptive in your comment. It's clear that you're setting all PCR values, but it's not clear _why_ you're setting them or _why_ you're setting the to this value.

DenisKarch replied:
I think it might be better to have that as a comment at the function signature.
We get all of the PCRs just so that we don't miss any data. Not all of them may have been used but the cost benefits to only getting some of the PCRs is kind of negligible.
Also I'm not sure if maybe this is a misunderstanding (because the naming is kind of terrible) but we're not setting the PCR values, were just setting a mask (the mask just isn't in the form of selected/not it's stored as a list of selected PCRs hence why we set the mask to range(24)).

ericchiang commented on attestation/attestation.go:
```diff
@@ -141,10 +140,52 @@ func CreateAIK(context *tspi.Context) ([]byte, []byte, error) {
+	for i := 0; i < 24; i++ {
+		bitmap[i] = i
+	}
+	err = pcrs.SetPCRs(bitmap)
```
nit, if you only return an error, you can do
```
if err := pcrs.SetPCRs(bitmap); err != nil {
    // handle error
}
```

ericchiang commented on attestation/attestation.go:
```diff
@@ -141,10 +140,52 @@ func CreateAIK(context *tspi.Context) ([]byte, []byte, error) {
+func GetQuote(context *tspi.Context, aikblob, nonce []byte) ([]byte, []byte, error) {
+	srk, err := context.LoadKeyByUUID(tspiconst.TSS_PS_TYPE_SYSTEM, tspi.TSS_UUID_SRK)
+	if err != nil {
+		return nil, nil, err
```
nit: fmt.Errorf for context here and in the next three returns

ericchiang commented on attestation/attestation.go:
```diff
@@ -141,10 +140,52 @@ func CreateAIK(context *tspi.Context) ([]byte, []byte, error) {
+	if err != nil {
+		return nil, nil, err
+	}
+	srkpolicy.SetSecret(tspiconst.TSS_SECRET_MODE_SHA1, wellKnown[:])
```
Add a comment explaining why you're using wellKnown here. Could this ever be a different value?

DenisKarch committed:
Made wellKnownSecret global

Made secret global (renamed for clarity)
Added comment explaining why we make our secret well known

ericchiang reviewed (approved):
lgtm, please squash and we can merge after

DenisKarch committed:
Added GetQuote

GetQuote consumes a nonce and the AIK blob and returns the quote (a SHA1
hash of the PCR values) plus a validation blob signed by the AIK and any
error that may have occurred.
